[SPARK-19554][UI,YARN] Allow SHS URL to be used for tracking in YARN RM.

vanzin opened the pull request:
Allow an application to use the History Server URL as the tracking
URL in the YARN RM, so there's still a link to the web UI somewhere
in YARN even if the driver's UI is disabled. This is useful, for
example, if an admin wants to disable the driver UI by default for
applications, since it's harder to secure it (since it involves non
trivial ssl certificate and auth management that admins may not want
to expose to user apps).

This needs to be opt-in, because of the way the YARN proxy works, so
a new configuration was added to enable the option.

The YARN RM will proxy requests to live AMs instead of redirecting
the client, so pages in the SHS UI will not render correctly since
they'll reference invalid paths in the RM UI. The proxy base support
in the SHS cannot be used since that would prevent direct access to
the SHS.

So, to solve this problem, for the feature to work end-to-end, a new
YARN-specific filter was added that detects whether the requests come
from the proxy and redirects the client appropriatly. The SHS admin has
to add this filter manually if they want the feature to work.

Tested with new unit test, and by running with the documented configuration
set in a test cluster. Also verified the driver UI is used when it's
enabled.

Commit:
[SPARK-19554][UI,YARN] Allow SHS URL to be used for tracking in YARN RM.

Allow an application to use the History Server URL as the tracking
URL in the YARN RM, so there's still a link to the web UI somewhere
in YARN even if the driver's UI is disabled. This is useful, for
example, if an admin wants to disable the driver UI by default for
applications, since it's harder to secure it (since it involves non
trivial ssl certificate and auth management that admins may not want
to expose to user apps).

This needs to be opt-in, because of the way the YARN proxy works, so
a new configuration was added to enable the option.

The YARN RM will proxy requests to live AMs instead of redirecting
the client, so pages in the SHS UI will not render correctly since
they'll reference invalid paths in the RM UI. The proxy base support
in the SHS cannot be used since that would prevent direct access to
the SHS.

So, to solve this problem, for the feature to work end-to-end, a new
YARN-specific filter was added that detects whether the requests come
from the proxy and redirects the client appropriatly. The SHS admin has
to add this filter manually if they want the feature to work.

Tested with new unit test, and by running with the documented configuration
set in a test cluster. Also verified the driver UI is used when it's
enabled.

SparkQA commented:
**[Test build #72967 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72967/testReport)** for PR 16946 at commit [`32b8dcd`](https://github.com/apache/spark/commit/32b8dcd9cc58a4f334f12a4f655228fe0383a332).

SparkQA commented:
**[Test build #72967 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72967/testReport)** for PR 16946 at commit [`32b8dcd`](https://github.com/apache/spark/commit/32b8dcd9cc58a4f334f12a4f655228fe0383a332).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class YarnProxyRedirectFilter extends Filter with Logging `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72967/
Test PASSed.

Commit:
Change constant name to be more readable.

vanzin commented:
@tgravescs pinging you since this is all YARN-side (doesn't really touch the UI).

SparkQA commented:
**[Test build #73007 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73007/testReport)** for PR 16946 at commit [`d675717`](https://github.com/apache/spark/commit/d675717a0c8459abf3d214a8a45c75b3de959941).

SparkQA commented:
**[Test build #73007 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73007/testReport)** for PR 16946 at commit [`d675717`](https://github.com/apache/spark/commit/d675717a0c8459abf3d214a8a45c75b3de959941).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73007/
Test PASSed.

vanzin commented:
trying to get some eyes on this: @squito

Commit:
Apply user filters to redirect handler.

Turn the redirect handler into a servlet, and install user filters,
so that they can be applied before redirection to the HTTPS port.
This can be used, for example, to perform redirection to the SHS
from the YARN RM proxy without having to deal with trust stores
on the YARN RM config.

SparkQA commented:
**[Test build #73242 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73242/testReport)** for PR 16946 at commit [`5aef8eb`](https://github.com/apache/spark/commit/5aef8eb569a5422be0c3c18906a04e173782c2ba).

tgravescs commented:
On vacation back next Monday and will review.

SparkQA commented:
**[Test build #73242 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73242/testReport)** for PR 16946 at commit [`5aef8eb`](https://github.com/apache/spark/commit/5aef8eb569a5422be0c3c18906a04e173782c2ba).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73242/
Test PASSed.

Commit:
Revert "Apply user filters to redirect handler."

It may not be the best idea to install auth filters on the
unencrypted connector, so don't do this and require admins
to properly set up trust stores in YARN instead. We can
add this as a new feature (with security properly accounted
for) later on.

SparkQA commented:
**[Test build #73287 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73287/testReport)** for PR 16946 at commit [`b380a85`](https://github.com/apache/spark/commit/b380a85b656c47aa53fda639036fa4f0e659bcc4).

SparkQA commented:
**[Test build #73287 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73287/testReport)** for PR 16946 at commit [`b380a85`](https://github.com/apache/spark/commit/b380a85b656c47aa53fda639036fa4f0e659bcc4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73287/
Test PASSed.

squito commented on docs/running-on-yarn.md:
```diff
@@ -604,3 +604,17 @@ spark.yarn.am.extraJavaOptions -Dsun.security.krb5.debug=true -Dsun.security.spn
+## Using the Spark History Server to replace the Spark Web UI
+
+It is possible to use the Spark History Server application page as the tracking URL for running
+applications in scenarios where it may be desired to disable the built-in application UI. Two steps
```
nit: first sentence reads a little funny.  maybe rephrase to:

It is possible to use the Spark History Server application page as the tracking URL for running
 applications where built-in application UI is disabled.  This may be desirable on secure clusters or to avoid the memory usage on the driver from the UI.

Up to you.  maybe doens't even need the second sentence.

squito reviewed (approved):
lgtm
very minor comments

Commit:
Feedback.

SparkQA commented:
**[Test build #73296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73296/testReport)** for PR 16946 at commit [`e0df6a5`](https://github.com/apache/spark/commit/e0df6a5bbaff4843d6d824e4403d4ff34bc5482f).

SparkQA commented:
**[Test build #73296 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73296/testReport)** for PR 16946 at commit [`e0df6a5`](https://github.com/apache/spark/commit/e0df6a5bbaff4843d6d824e4403d4ff34bc5482f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73296/
Test PASSed.

vanzin commented:
Merging to master.

@tgravescs we can address any feedback you might have when you're back.
